odoo/addons/base/models/res_users.py

```diff
@@ -1124,20 +1124,30 @@ def create(self, vals_list):
                 values['groups_id'] = type(self).groups_id.convert_to_write(gs, user)
         return super(UsersImplied, self).create(vals_list)
 
+    def _clear_demoted_user_access(self, values, users_before):
+        portal_users_to_clear = self.env['res.users']
+        for user in self:
+            if not user.has_group('base.group_user') and user in users_before:
+                portal_users_to_clear |= user
+        if portal_users_to_clear:
+            vals = {'groups_id': [Command.clear()] + values['groups_id']}
+            super(UsersImplied, portal_users_to_clear).write(vals)
+        return portal_users_to_clear
+
     def write(self, values):
         users_before = self.filtered(lambda u: u.has_group('base.group_user'))
         res = super(UsersImplied, self).write(values)
         if values.get('groups_id'):
-            # add implied groups for all users
+            self._clear_demoted_user_access(values, users_before)
+            users_batch = defaultdict(self.browse)
             for user in self:
-                if not user.has_group('base.group_user') and user in users_before:
-                    # if we demoted a user, we strip him of all its previous privileges
-                    # (but we should not do it if we are simply adding a technical group to a portal user)
-                    vals = {'groups_id': [Command.clear()] + values['groups_id']}
-                    super(UsersImplied, user).write(vals)
+                users_batch[frozenset(user.groups_id.ids)] |= user
```

Comment:
Recordsets are hashable
```suggestion
                users_batch[user.groups_id] |= user
```
And by the way, the new key can simplify the loop below.

Reply:
I used to avoid recordset as key, because I used to get some weird result with this (I guess because recordset is related with cursor that can be closed?). But it might not be applicable here as we are using all data without holding it for long time.